component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java

```diff
@@ -0,0 +1,186 @@
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.oauth.common.OAuth2ErrorCodes;
+import org.wso2.carbon.identity.oauth.common.exception.InvalidOAuthClientException;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.bean.OAuthClientAuthnContext;
+import org.wso2.carbon.identity.oauth2.client.authentication.BasicAuthClientAuthenticator;
+import org.wso2.carbon.identity.oauth2.client.authentication.OAuthClientAuthnException;
+import org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret.util.MutualTLSUtil;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import javax.servlet.http.HttpServletRequest;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret.util.MutualTLSUtil.JAVAX_SERVLET_REQUEST_CERTIFICATE;
+
+/**
+ * This class is an enhancement for BasicAuthClientAuthenticator which validates the client certificate
+ * in addition to the client id and secret authentication
+ *
+ */
+public class MutualTLSWithIdSecretAuthenticator extends BasicAuthClientAuthenticator {
+
+    private static Log log = LogFactory.getLog(MutualTLSWithIdSecretAuthenticator.class);
+
+    public boolean authenticateClient(HttpServletRequest request, Map<String, List> bodyParams,
+                                      OAuthClientAuthnContext oAuthClientAuthnContext)
+            throws OAuthClientAuthnException {
+
+        if (!super.authenticateClient(request, bodyParams, oAuthClientAuthnContext)) {
+            return false;
+        }
+
+        if (StringUtils.isEmpty(oAuthClientAuthnContext.getClientId())) {
+            oAuthClientAuthnContext.setClientId(this.getClientId(request, bodyParams, oAuthClientAuthnContext));
+        }
+
+        try {
+
+            String tenantDomain = OAuth2Util.getTenantDomainOfOauthApp(oAuthClientAuthnContext.getClientId());
+            X509Certificate registeredCert = null;
+            try {
+                registeredCert = (X509Certificate) OAuth2Util
+                        .getX509CertOfOAuthApp(oAuthClientAuthnContext.getClientId(), tenantDomain);
+            } catch (IdentityOAuth2Exception e) {
+                if (e.getCause() instanceof CertificateException) {
+                    throw e;
+                } else {
+                    // This means certificate is not configured in service provider. In that case basic authentication
+                    // would be performed
+                    return super.authenticateClient(request, bodyParams, oAuthClientAuthnContext);
```

Comment:
Do we need to re-authenticate here ?  We have already passed basic authentication